control-plane/cni/main.go

```diff
@@ -0,0 +1,223 @@
+	if err := types.LoadArgs(args.Args, &cniArgs); err != nil {
+		return err
+	}
+
+	podNamespace := string(cniArgs.K8S_POD_NAMESPACE)
+	podName := string(cniArgs.K8S_POD_NAME)
+
+	// We only run in a pod
+	if podNamespace == "" && podName == "" {
+		return fmt.Errorf("not running in a pod, namespace and pod should have values")
+	}
+
+	logPrefix := fmt.Sprintf("%s/%s", podNamespace, podName)
+	logger := hclog.New(&hclog.LoggerOptions{
+		Name:  logPrefix,
+		Level: hclog.LevelFromString(cfg.LogLevel),
+	})
+
+	// Check to see if the plugin is a chained plugin
+	if cfg.PrevResult == nil {
+		return fmt.Errorf("must be called as chained plugin")
+	}
+
+	logger.Debug("consul-cni plugin config", "config", cfg)
+	// Convert the PrevResult to a concrete Result type that can be modified. The CNI standard says
+	// that the previous result needs to be passed onto the next plugin
+	prevResult, err := current.GetResult(cfg.PrevResult)
+	if err != nil {
+		return fmt.Errorf("failed to convert prevResult: %v", err)
+	}
+
+	if len(prevResult.IPs) == 0 {
+		return fmt.Errorf("got no container IPs")
+	}
+
+	// Pass the prevResult through this plugin to the next one
+	result := prevResult
+	logger.Debug("consul-cni previous result", "result", result)
+
+	ctx := context.Background()
+	restConfig, err := clientcmd.BuildConfigFromFlags("", filepath.Join(cfg.CNINetDir, cfg.Kubeconfig))
+	if err != nil {
+		return fmt.Errorf("could not get rest config from kubernetes api: %s", err)
+	}
+
+	client, err := kubernetes.NewForConfig(restConfig)
+	if err != nil {
+		return fmt.Errorf("Error initializing Kubernetes client: %s", err)
+	}
+
+	pod, err := client.CoreV1().Pods(podNamespace).Get(ctx, podName, metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+
+	// TODO: Add tests for all of the logic below
+	// TODO: Remove has hasBeenInjected and instead look for consul.hashicorp.com/cni-proxy-config annotation
+	// TODO: Add wait and timeout for annotations to show up
+	if hasBeenInjected(*pod) {
+		// If everything is good, add an annotation to the pod
+		// TODO: Remove this as it is just a stub to prove that we can do kubernetes things with the plugin
+		annotations := map[string]string{
+			keyCNIStatus: "true",
+		}
+		pod.SetAnnotations(annotations)
+		_, err = client.CoreV1().Pods(podNamespace).Update(context.TODO(), pod, metav1.UpdateOptions{})
+		if err != nil {
+			return err
+		}
+	} else {
+		logger.Debug("skipping traffic redirect on un-injected pod")
+	}
+
+	// TODO: Get transparent proxy annotations and merge with proxy config
+	// TODO: Redirect traffic :)
+
+	// Pass through the result for the next plugin
+	return types.PrintResult(result, cfg.CNIVersion)
+}
+
+// cmdDel is called for DELETE requests.
+func cmdDel(args *skel.CmdArgs) error {
+	conf, err := parseConfig(args.StdinData)
+	if err != nil {
+		return err
+	}
+	_ = conf
+
+	// Nothing for consul-cni plugin to do as everything is removed once the pod is gone
+	return nil
+}
+
+func main() {
```

Comment:
nit: ordering with these function is weird, can `main()` be below `cmdCheck()`?

Reply:
Done.